Provide initial README

ctubbsii opened the pull request:
Initial readme for review

ctubbsii committed:
Provide initial README

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+[![Build Status][ti]][tl] [![Apache License][li]][ll] [![Maven Central][mi]][ml] [![Javadoc][ji]][jl]
+
+**Apache Fluo Bytes is an API project with the goal of providing an extremely
+stable library for handling bytes, suitable for use in [Apache Fluo][fluo] and
```
A few things I found unclear.

  * What is an `API project`?
  * What does `stable library` mean?

Below is my attempt to clear those things up.  My main goal was to state its a library with a stable API.

```
Apache Fluo Bytes is a project with the goal of providing a bytes library with an extremely stable API that is suitable for use  in other projects' APIs.
```

ctubbsii replied:
I didn't see your exact wording above, because it didn't show correctly when I saw this comment in my email, but I pushed a change which basically applies the same fix, based on your feedback above the blockquote section.

keith-turner replied:
I like your wording better.  One thing I didn't like in my wording was `bytes library`.

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+## Features and Goals
+
+This project aims to fill a void in Java, by providing convenient objects to
+represent a string of bytes and associated utility classes for situations when
```
could say sequence of bytes to since in java String is so strongly associated with char

ctubbsii replied:
Fixed.

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+
+This project aims to fill a void in Java, by providing convenient objects to
+represent a string of bytes and associated utility classes for situations when
+a raw byte array is not appropriate.
```
We could  point out the problems with raw byte array.
 * Mutable
 * no equals
* no hash code
* not comparable

ctubbsii replied:
I did try to write more on this initially, but everything I came up with was too long and seemed to drift off topic.

I boiled it down to what you see here, mentioning this library is useful when byte arrays aren't appropriate, and then immediately following that statement by some of the features that make it so, naming all of the things you list here. I don't know that I'd really want more details in the README than that. I certainly don't want to repeat a blog post, or make the README a big wall of text that provides so many details, it's hard to get the gist.

But, if you have specific suggestions which can improve this, I'm happy to work it in. I, however, am near my editing limit on this subject. :)

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+
+* Using [Semantic Versioning 2.0.0][semver] to make strong declarations about
+  backwards-compatibility
+* Strongly avoid breaking changes (avoid major version bumps), so that projects
```
It will be so nice in Java 9 when the deprecated annotation has the forRemoval parameter.

ctubbsii replied:
Ugh, I hate that new "feature" to Java 9, especially since it defaults to "false". Personally, I think all deprecations are subject to removal. I really don't want to start putting "forRemoval = true" in all my deprecation annotations just so users don't think "this isn't going away".

I'd rather they'd have created a new annotation for things which are expected to remain, but should not be used by users, which generated similar compiler warnings as Deprecated.

This one is really going to cause confusion with defaulting to false. A lot of them are going to be set to false simply because nobody's explicitly setting them to true. Users are going to be disappointed when they disappear.

keith-turner replied:
> I'd rather they'd have created a new annotation for things which are expected to remain, but should not be used by users, which generated similar compiler warnings as Deprecated.

That would have been better.

keith-turner replied:
>  I think all deprecations are subject to removal

I think this depends on the project.  I realized I have never looked at the javadoc for the annotation, it says the following.

```
A program element annotated @Deprecated is one that programmers are discouraged from using, typically because it is dangerous, or because a better alternative exists.  Compilers warn when a deprecated program element is used or overridden in non-deprecated code. 
```

Removal is not mentioned, so the new parameter will mesh well with the current javadoc.

ctubbsii replied:
Yeah, removal is not explicitly mentioned, but if it's "dangerous" or if a "better alternative exists", it seems appropriate to assume it is "subject to" removal, even if removal is deferred indefinitely because of other factors (stability). Part of the problem with the new flag is that "subject to" is not well-defined. I hear "subject to" and think "could possibly be, at the determination of the developers"... but that's basically *always the case*.

ctubbsii committed:
Improve wording of README

Apply feedback from code review:

* Avoid "string of bytes", prefer "sequence of bytes" so as not to
invoke the Java concept of "String", which is associated with chars
* Improve basic description of project as a library with a stable API

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+* Strongly avoid breaking changes (avoid major version bumps), so that projects
+  can converge on the latest version required by their code and that of their
+  dependencies, without risk of incompatibility
+* No runtime dependencies to use the library, to eliminate any potential
```
`No runtime dependencies to use the library`  this seems a bit off.  To use this library a runtime dep on this library will be needed.

keith-turner commented on README.md:
```diff
@@ -0,0 +1,102 @@
+* Using [Semantic Versioning 2.0.0][semver] to make strong declarations about
+  backwards-compatibility
+* Strongly avoid breaking changes (avoid major version bumps), so that projects
+  can converge on the latest version required by their code and that of their
```
When I read `the latest version required`  I found myself wondering if `the latest version of Fluo Bytes required` would be better.  I like it because its more precise. I don't like that it adds words I think people could infer (but not completely sure because it all makes perfect sense to me).  I am not advocating for this change, it was just a nagging thought I had when reading this.

This document frequently uses `this project`.  When thinking about the above I started mulling over whether replacing `this project` with `Fluo Bytes` would be a good thing.  I have not come to any conclusion.

ctubbsii replied:
Made some additional tweaks to the wording to address these, but avoided including "Fluo Bytes" redundantly everywhere.

ctubbsii committed:
Additional README wording improvements
